### PR TITLE
Consolidate bundle and BD controller code

### LINKED ISSUE #575: DRY up copy/pasted provisioner reconciler code
We currently have a large amount of copy/pasted code in our various in-tree provisioners. We have three bundle provisioners: plain, registry, and helm. And we have two BD provisioners: plain and helm.

A very large majority of the code in each of these provisioners is identical, and the only differences are in:
- how the bundle content is validated and converted before storing
- how the bundle content is validated and rendered into a helm chart before being applied to the cluster.

We should look into introducing a few interfaces to capture these implementation-specific differences along with generic bundle and BD reconcilers that use these interfaces such that the copy/pasted code can be shared and DRY-ed up.

### PR DESCRIPTION
**Pull request by joelanford**
Closes #575

### CODE REVIEW

**Comment by joelanford** [comment]
New libraries are:
- `internal/provisioner/bundle`
- `internal/provisioner/bundledeployment`

Interface implementations are still under `internal/provisioner/<format>`, but code in those directories is now limited to _just_ the interface implementations.

Main commands were updated to build the reconcilers with the new libraries and format-specific implementations.

File: cmd/core/main.go [hunk 1]
```diff
@@ -205,37 +207,44 @@ func main() {
...
-		Scheme:     mgr.GetScheme(),
-		Storage:    bundleStorage,
-		Finalizers: bundleFinalizers,
-		Unpacker:   unpacker,
-	}).SetupWithManager(mgr); err != nil {
-		setupLog.Error(err, "unable to create controller", "controller", rukpakv1alpha1.BundleKind)
+	if err := bundle.SetupProvisioner(mgr, append(
+		commonBundleProvisionerOptions,
+		bundle.WithProvisionerID(registry.ProvisionerID),
+		bundle.WithHandler(bundle.HandlerFunc(registry.HandleBundle)),
```
Comment (akihikokuroda) [inline cmd/core/main.go:236]: It took some time for me to understand this line.  Is this a popular way to pass an interface as an argument?

File: internal/provisioner/bundledeployment/bundledeployment.go [hunk 2]
```diff
@@ -409,19 +471,19 @@ func isResourceNotFoundErr(err error) bool {
 }
 
 // SetupWithManager sets up the controller with the Manager.
-func (r *BundleDeploymentReconciler) SetupWithManager(mgr ctrl.Manager) error {
+func (p *bundledeploymentProvisioner) SetupWithManager(mgr ctrl.Manager) error {
```
Comment (akihikokuroda) [inline internal/provisioner/bundledeployment/bundledeployment.go:474]: Is this still be used?

File: internal/provisioner/registry/types.go [hunk 3]
```diff
@@ -0,0 +1,28 @@
+package registry
```
Comment (akihikokuroda) [inline internal/provisioner/registry/types.go:1]: Should this file be renamed to retistry.go for consistency?

**Comment by akihikokuroda** [comment]
This looks good.  I have some minor comments.

File: cmd/core/main.go (hunk 1 above)
Comment (joelanford) [reply cmd/core/main.go:236]: This seems common with single-method interfaces. I think the most well-known one is `http.Handler`/`http.HandlerFunc`.

File: internal/provisioner/bundle/bundle.go [hunk 4]
```diff
@@ -1,59 +1,125 @@
...
+
+func (p *bundleProvisioner) setDefaults() {
+	if p.handler == nil {
+		p.handler = HandlerFunc(func(_ context.Context, fsys fs.FS, _ *rukpakv1alpha1.Bundle) (fs.FS, error) { return fsys, nil })
+	}
+}
+
+func (p *bundleProvisioner) validateConfig() error {
+	errs := []error{}
+	if p.handler == nil {
```
[the remark this answers is not in the document]
Comment (joelanford) [reply internal/provisioner/bundle/bundle.go:97]: I was trying to separate concerns a little bit and also have defensive code in the validate function (i.e. don't assume that `setDefaults` exists or ensures handler is always set).

I can combine into a single function if there's consensus.

File: internal/provisioner/bundledeployment/bundledeployment.go (hunk 2 above)
Comment (joelanford) [reply internal/provisioner/bundledeployment/bundledeployment.go:474]: Don't think so. Good catch.

**Review by akihikokuroda — APPROVED** [review]

**Comment by tylerslaton** [comment]
The code here looks good and it is blocking some work that @akihikokuroda is doing. I'm going to go ahead and hit the merge button on @joelanford's behalf as he is on PTO for awhile.

**Pull request merged**